Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1

```diff
@@ -0,0 +1,370 @@
+using module Microsoft.AVS.Management
+
+<#
+    .SYNOPSIS
+     Creates a new vVol datastore and mounts to a VMware cluster.
+
+    .PARAMETER ClusterName
+     Cluster name
+
+    .PARAMETER DatastoreName
+     Datastore name
+
+    .PARAMETER ScId
+     Storage container ID of device used to create a new vVol datastore
+
+    .EXAMPLE
+     New-VVolDatastore -ClusterName "myCluster" -DatastoreName "myDatastore" -ScId $ScId
+
+    .INPUTS
+     vCenter cluster name, datastore name, and storage container ID
+
+    .OUTPUTS
+     None.
+#>
+function New-VvolDatastore {
+    [CmdletBinding()]
+    [AVSAttribute(10, UpdatesSDDC = $false)]
+    Param (
+        [Parameter(
+            Mandatory = $true,
+            HelpMessage = 'Cluster name in vCenter')]
+        [ValidateNotNull()]
+        [String]
+        $ClusterName,
+
+        [Parameter(
+            Mandatory = $true,
+            HelpMessage = 'Name of vVol datastore to be created in vCenter')]
+        [ValidateNotNull()]
+        [String]
+        $DatastoreName,
+
+        [Parameter(
+            Mandatory = $true,
+            HelpMessage = 'Storage container ID of device used to create a new vVol datastore')]
+        [ValidateNotNull()]
+        [String]
+        $ScId
+    )
+
+    $Cluster = Get-Cluster -Name $ClusterName -ErrorAction Ignore
+    if (-not $Cluster) {
+        throw "Cluster $ClusterName does not exist."
+    }
+
+    $Datastore = Get-Datastore -Name $DatastoreName -ErrorAction Ignore
+    if ($Datastore) {
+        throw "Unable to create a datastore. Datastore '$DatastoreName' already exists."
+    }
+    $VMHosts = $Cluster | Get-VMHost
+    # We need to loop through Esxi to mount the datastore to all of hosts
+    foreach ($Esxi in $VMHosts) {
+        # Create a new vVol datastore with the specified size and rescan storage
+        $datastoreSystem = Get-View -Id $Esxi.ExtensionData.ConfigManager.DatastoreSystem
+        $spec = New-Object VMware.Vim.HostDatastoreSystemVvolDatastoreSpec
+        $spec.Name = $datastoreName
+        $spec.ScId = $scId
+        $Datastore = $esxi | Get-Datastore | Where-Object { $_.Type -eq "VVol" } | Where-Object { $_.ExtensionData.Info.VVolds.Scid -eq $scId }
+        Write-Host "Mounting datastore $DatastoreName to host $($Esxi.Name)..."
+        $datastoreSystem.CreateVvolDatastore($spec)
+    }
+}
+
+
+<#
+    .SYNOPSIS
+     Removes a vVol datastore from a VMware cluster.
+
+    .PARAMETER ClusterName
+     Cluster name
+
+    .PARAMETER DatastoreName
+     Datastore name
+
+    .EXAMPLE
+     Remove-VVolDatastore -ClusterName "myCluster" -DatastoreName "myDatastore"
+
+    .INPUTS
+     vCenter cluster name, datastore name
+
+    .OUTPUTS
+     None.
+#>
+function Remove-VvolDatastore {
```

Comment:
Could we check as a precondition that there are no VMs using this datastore?

Reply:
Done